packages/client/src/iterators-utils.ts

```diff
@@ -0,0 +1,74 @@
+import { EventEmitter, on } from 'node:events';
+
+/**
+ * Return an async iterable that transforms items from a source iterable by mapping each item
+ * through a mapping function.
+ *
+ * @param source the source async iterable
+ * @param mapFn a mapping function to apply on every item of the source iterable
+ */
+export async function* mapAsyncIterable<A, B>(
+  source: AsyncIterable<A>,
+  mapFn: (val: A) => Promise<B>
+): AsyncIterable<B> {
+  for await (const x of source) {
+    yield mapFn(x);
+  }
+}
+
+/**
+ * Return an async iterable that transforms items from a source iterable by concurrently mapping
+ * each item through a mapping function.
+ *
+ * Mapped items are returned by the resulting iterator in the order they complete processing,
+ * not the order in which the corresponding source items were obtained from the source iterator.
+ *
+ * @param source the source async iterable
+ * @param mapFn a mapping function to apply on every item of the source iterable
+ * @param concurrency maximum number of items to map concurrently
+ */
+export async function* raceMapAsyncIterable<A, B>(
+  source: AsyncIterable<A>,
+  mapFn: (val: A) => Promise<B>,
+  concurrency: number
+): AsyncIterable<B> {
+  const emiter = new EventEmitter();
+  const sourceIterator = source[Symbol.asyncIterator]();
+  let sourceIteratorDone = false;
+  let pendingPromisesCount = 0;
+  let pendingResultsCount = 0;
+
+  async function maybeStartTasks() {
+    while (!sourceIteratorDone && pendingPromisesCount + pendingResultsCount < concurrency) {
+      const val = await sourceIterator.next();
+      if (val.done) {
+        sourceIteratorDone = true;
+        return pendingPromisesCount > 0;
+      }
+
+      pendingPromisesCount++;
+      (async () => {
+        try {
+          emiter.emit('result', await mapFn(val.value));
+          pendingResultsCount++;
+        } finally {
+          pendingPromisesCount--;
+        }
+        return maybeStartTasks();
+      })().catch((e) => emiter.emit('error', e));
+    }
+
+    return !sourceIteratorDone || pendingPromisesCount > 0;
+  }
+
+  // The listener must be registered before we start emiting events
+  // through the emiter; otherwise, these events get silently discarded.
+  const emiterEventsIterable = on(emiter, 'result');
+
+  if (!(await maybeStartTasks())) return;
```

Comment:
I would do this a bit differently.

Spawn `concurrency` coroutines each running:

```
for (;;) {
  await sourceIterator.next();
  ...
}
```

And create an `AbortController` (or simple promise) that you can use to signal the coroutines to stop on first error (in your code they keep consuming the iterator).

This way you can get rid of the spawning logic and keeping track of `sourceIteratorDone, pendingPromisesCount, pendingResultsCount`.

My version looks roughly like:

```ts
  const emitter = new EventEmitter();
  const controller = new AbortController();
  const emitterEventsIterable = on(emitter, 'result', { signal: controller.signal });
  const emitterError = once(emitter, 'not-an-event'); // so we can break on error.

  const consume = async () => {
    for (;;) {
      const val = await Promise.race([sourceIterator.next(), emitterError]);
      if (val.done) return;

      emitter.emit('result', await mapFn(val.value));
    }
  };

  const coroutines = Array(concurrency)
    .fill(consume)
    .map((f) => f());
  Promise.all(coroutines).then(
    () => controller.abort(),
    (err) => emitter.emit('error', err)
  );

  try {
    for await (const [res] of emitterEventsIterable) {
      yield res;
    }
  } catch (err) {
    if (err.name === 'AbortError') {
      return;
    }
    throw err;
  }

```

Reply:
Yeah, I had considered something similar to this at some point, and it definitely look simpler...

My concern however is that I think we really need backpressure control (which, at this point, is provided by the `pendingPromisesCount + pendingResultsCount < concurrency` condition), at least for "fail fast" scenarios. Otherwise, the iterator risk consuming a very large quantity of data from the server, only to discard that data at the first error...

But let me think about that a little more, it might not be that difficult to introduce backpressure control in the pattern you suggest.

Reply:
I see your concern, definitely a valid one.
You could implement a semaphore and release the permits on consumption (which is more or less what you did here).

Reply:
Done

Reply:
I saw the last commit.
LGTM with the exception that I'd move the semaphore implementation out of this function for better readability.
Looks like you need `Queue` and `Semaphore` classes.

Reply:
> I'd move the semaphore implementation out of this function for better readability. Looks like you need Queue and Semaphore classes.

Will do at some later point.